### ISSUE TITLE
Page numbers in PDF bookmarks deviate from ToC.

### ISSUE THREAD

**Issue by ascherer** [ascherer/web#25]
Consult project [latex-tds](https://github.com/oberdiek/latex-tds) (also on [CTAN](https://ctan.org/pkg/latex-tds)) for ideas. In 2020 I extracted enough code from [knuth.drv](https://github.com/oberdiek/latex-tds/blob/master/tex/knuth.drv) (contained in both `latex-tds.tds.zip` and `source.tds.zip`) into [pdfwebtocfront.tex](https://github.com/ascherer/cwebbin/blob/master/texinputs/pdfwebtocfront.tex) to make things work. Heiko Oberdiek's full code manages to manipulate page numbers in PDF files according to TeX's counter.

This affects at least parts of
* `cwebmac.tex`, `cwebacromac.tex`
* `pwebmac.tex`
* `pdfwebtocfront.tex`
* `makeall` (SED changes)

**Comment by ascherer** [comment]
Consult [latex-tds](https://ctan.org/pkg/latex-tds) for ideas. Heiko Oberdieck manages to manipulate page numbers according to TeX's counter.

**Comment by ascherer** [comment]
Here is a direct comparison between Heiko's [gftodvi.pdf from 2014](https://github.com/ascherer/web/files/15406728/gftodvi.pdf) and my [gftodvi.pdf from 2024](https://github.com/ascherer/web/files/15406735/gftodvi.pdf).

**Comment by ascherer** [comment]
Extract `\makepagelabels` from `knuth.drv`.

**Comment by ascherer** [comment]
This code adapts `cwebmac.tex` upstream and works with `pdfwebtocfront.tex`:
```
From b41f30c0a1d66bec3420354c3cc82b887ea29dd6 Mon Sep 17 00:00:00 2001
From: Andreas Scherer <andreas_github@freenet.de>
Date: Fri, 24 May 2024 11:46:36 +0200
Subject: [PATCH] Match page numbers in PDF bookmarks with ToC.

---
 cwebmac.tex | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/cwebmac.tex b/cwebmac.tex
index a59f15b..c0116a1 100644
--- a/cwebmac.tex
+++ b/cwebmac.tex
@@ -276,8 +276,16 @@
 \def\title{\expandafter\uppercase\expandafter{\jobname}}
 \def\topofcontents{\centerline{\titlefont\title}\vskip.7in
   \vfill} % this material will start the table of contents page
-\def\startpdf{\ifpdf \ifpdflua\pdfcatalog{/PageMode /UseOutlines}
-  \else {\special{pdf: docview << /PageMode /UseOutlines >>}}\fi\fi}
+\def\startpdf{\ifpdf \def\pagemode{/PageMode /UseOutlines}
+  \ifnum\contentspagenumber=0 \advance \pageno by -1
+    \def\pagelabels{/PageLabels << /Nums [ 0 << /S/D/St 1 >>
+      \the\pageno << /P(Contents) >> ] >>}
+  \else \twodigits=\pageno \pageno=\contentspagenumber
+    \advance \pageno by 1 \advance \twodigits by -\pageno
+    \def\pagelabels{/PageLabels << /Nums [ 0 << /S/D/St \the\pageno >>
+      \the\twodigits << /S/D/St \contentspagenumber >> ] >>} \fi
+  \ifpdflua\pdfcatalog{\pagemode\space\pagelabels}
+  \else \special{pdf: docview << \pagemode\space\pagelabels >>}\fi\fi}
 \def\botofcontents{\vfill
   \centerline{\covernote}} % this material will end the table of contents page
 \def\covernote{}
@@ -383,12 +391,12 @@
 % \ifodd\pageno\else\titletrue\null\vfill\eject\fi % for duplex printers
   \rightskip 0pt \hyphenpenalty 50 \tolerance 200
   \setpage \output={\normaloutput\page\lheader\rheader}
-  \titletrue % prepare to output the table of contents
+  \startpdf \titletrue % prepare to output the table of contents
   \pageno=\contentspagenumber
   \def\grouptitle{TABLE OF CONTENTS}
   \message{Table of contents:}
   \ifhint\HINThome\fi% Mark the Table of contents as home page
-  \topofcontents \startpdf
+  \topofcontents
   \line{\hfil Section\ifhint\else\hbox to3em{\hss Page}\fi}% No Page in HINT
   \let\ZZ=\contentsline
   \readcontents\relax % read the contents info
-- 
2.45.1
```

**Comment by ascherer** [comment]
Released in **pwebmac 4.12** and in **TeX Live**.
